File: client/spec/app/views/login_spec.js
//= require "../../hyperarchy_spec_helper"

Screw.Unit(function(c) { with(c) {
  describe("Views.Login", function() {
    useFakeApplicationController();

    var view;
    before(function() {
      view = Views.Login.toView();
    });


    describe("when #loginSubmit is clicked", function() {
      before(function() {
        view.find('#emailAddress').val("cobham@gmail.com");
        view.find('#password').val("spectrum");

        expect(Server.posts).to(beEmpty);
        view.find('#loginSubmit').click();

      });

      it("posts the form field values to /login after underscoring the keys", function() {
        expect(Server.posts.length).to(eq, 1);
        expect(Server.lastPost.url).to(eq, "/login");
        expect(Server.lastPost.data).to(equal, {
          email_address: "cobham@gmail.com",
          password: "spectrum"
        });
      });


      context("when the result is successful", function() {
        it("calls Application.currentUserIdEstablished with the User's id and navigates to the organization view", function() {

          mock(Application, 'currentUserIdEstablished');
          mock(History, 'load', function() {
            expect(Application.currentUserIdEstablished).to(haveBeenCalled, withArgs("billy"));
          });
          Server.lastPost.simulateSuccess({
            currentUserId: "billy"
          });
          expect(History.load).to(haveBeenCalled, withArgs("organizations"));
        });
      })

      context("when the result is unsuccessful", function() {
        it("displays the error and does not set the Application.currentUserId or navigate to #elections", function() {
          mock(History, 'load');
          Server.lastPost.simulateFailure({
            errors: {
              password: "Your password did not match the given email address."
            }
          });
          expect(History.load).toNot(haveBeenCalled);
          expect(view.html()).to(match, "Your password did not match the given email address.");
        });
      })

    });
  });
}});

File: server/spec/app/resources/login_spec.rb
require File.expand_path("#{File.dirname(__FILE__)}/../../hyperarchy_spec_helper")

module Resources
  describe Login, :type => :resource do
    describe "#post" do
      attr_reader :user

      before do
        @user = User.create(:email_address => "billy@example.com", :password => "spectrum")
      end

      context "if a User with the given email address exists" do
        context "if the given password matches that User" do
          it "sets the current_user and returns a successful ajax response with the current_user_id" do
            current_user.should be_nil
            post "/login", :email_address => user.email_address, :password => "spectrum"
            current_user.should == user

            response.body_from_json.should == {
              "successful" => true,
              "data" => {
                "current_user_id" => user.id  
              }
            }
          end
        end

        context "if the given password does NOT match that User" do
          it "does not set the current_user and returns an unsuccessful ajax response with errors on password" do
            current_user.should be_nil
            post "/login", :email_address => user.email_address, :password => "incorrectpassword"
            current_user.should be_nil

            response.body_from_json.should == {
              "successful" => false,
              "data" => {
                "errors" => {
                  "password" => "Incorrect password for the given email address."
                }
              }
            }
          end
        end
      end

      context "if no User with the given email address exists" do
        it "does not set the current_user and returns an unsuccessful ajax response with errors on email_address" do
          current_user.should be_nil
          post "/login", :email_address => "bogus@example.com", :password => "spectrum"
          current_user.should be_nil

          response.body_from_json.should == {
            "successful" => false,
            "data" => {
              "errors" => {
                "email_address" => "No user exists with this email address."
              }
            }
          }
        end
      end
    end
  end
end